src/sentry/ratelimits/cardinality.py

```diff
@@ -0,0 +1,139 @@
+"""
+A kind of limiter that limits set cardinality instead of a rate/count.
+
+The high-level concepts are very similar to `sentry.ratelimits.sliding_windows`.
+
+Instead of passing in numbers and getting back smaller numbers, however, the
+user passes in a set and gets back a smaller set. Set elements that have
+already been observed in any quota's window are "for free" and will not
+count towards any quota.
+
+The implementation hasn't been finalized yet, but we expect that under the hood
+this cardinality limiter will be more expensive to operate than a simple rate
+limiter, as it needs to keep track of already-seen set elements. The memory
+usage in Redis will most likely be proportional to the set size.
+
+This kind of limiter does not support prefix overrides, which practically means
+that there can only be a per-org or a global limit, not both at once.
+"""
+
+
+import time
+from dataclasses import dataclass
+from typing import Any, Collection, Iterator, Optional, Protocol, Sequence, Tuple, TypeVar
+
+from sentry.utils.services import Service
+
+Hash = str
```

Comment:
Why not int ? I think we should never need to convert the hash into a string.

Reply:
str seemed easier to use but I changed it

Reply:
Though we will need to convert to int before writing or reading to/from Redis. All our esitmates are based on 16/32 bit hashes. Strings will be much bigger, which means we will have to revisit the estimates and also waste space.